Update evaluate_loop.sh

UFO-101 opened the pull request:
Update `evaluate_loop.sh` so that victim and adversary names are set to the same format as in training. Add more options including `VICTIMS_LIST` which enables running against multiple victims.

Ticket: https://www.notion.so/chaiberkeley/evaluate_loop-sh-change-victim-and-adversary-name-to-include-metadata-and-add-support-for-evaluati-70d91ac930e043bbac4f85e4765027c1

UFO-101 committed:
Update evaluate_loop.sh so that victim and adversary names are set to
the same format as in training. Add more options including VICTIMS_LIST
which enables running against multiple victims.

Ticket: https://www.notion.so/chaiberkeley/evaluate_loop-sh-change-victim-and-adversary-name-to-include-metadata-and-add-support-for-evaluati-70d91ac930e043bbac4f85e4765027c1

tomtseng commented on cpp/evaluate_loop.sh:
```diff
@@ -1,20 +1,25 @@
 #!/bin/bash -eu
-if [[ $# -lt 1 || $# -gt 2 ]]
+if [[ $# -lt 1 || $# -gt 5 ]]
```
is `-gt 5` correct? looks like we have substantially more than that many arguments

AdamGleave replied:
:+1: looks like we can have up to 7 arguments?

tomtseng commented on cpp/evaluate_loop.sh:
```diff
@@ -1,20 +1,25 @@
     echo "Currently expects to be run from within the 'cpp' directory of the KataGo repo."
     echo "BASE_DIR is the root of the training run, containing selfplay data, models and related directories."
+    echo "VICTIM_LIST is a space or comma-separated list of models in VICTIM_DIR. If empty the most recent model will be used."
     echo "PREDICTOR_DIR is the path containing predictor models, if applicable."
```
Kinda inconvenient to specify the PREDICTOR_DIR now since we'd need to specify all the other positional arguments as well. We may want to convert these from positional arguments to flags (e.g., see how `go_attack/kubernetes/launch-job.sh` handles arguments).

AdamGleave replied:
:+1: flags seem better.

UFO-101 replied:
Switched to flags 👍

tomtseng commented on cpp/evaluate_loop.sh:
```diff
@@ -1,20 +1,25 @@
+if [[ $# -lt 1 || $# -gt 5 ]]
 then
-    echo "Usage: $0 BASE_DIR [PREDICTOR_DIR]"
+    echo "Usage: $0 BASE_DIR OUTPUT_DIR [KATAGO_BIN='/engines/KataGo-custom/cpp/katago GO_ATTACK_ROOT='/go_attack' VICTIM_LIST='' VICTIM_DIR='BASE_DIR/victims'  PREDICTOR_DIR='']"
```
When do we ever want to set KATAGO_BIN or GO_ATTACK_ROOT to anything else? Is it so that we can run the script outside the `goattack:cpp` container and make it easier to debug depending on the developer's setup?

AdamGleave replied:
Personally fine with having extra arguments so long as the defaults are fine, being able to run things from outside container can be quite helpful for development, but some documentation on this could help (e.g. adding usage in the echo statements below)

UFO-101 replied:
Using the VSCode Dev Containers setup described [here](https://github.com/HumanCompatibleAI/go_attack/pull/72), you want to have different directories because the default compilation directory is `${workspaceFolder}/build/katago` (when building with the C++ extension) and the repo root is `workspace`.

UFO-101 replied:
Added more info to flags.

tomtseng commented on cpp/evaluate_loop.sh:
```diff
@@ -26,52 +31,63 @@ do
+        VICTIM_LIST=$(ls -Art "$VICTIMS_DIR" | tail -n 1)
+    fi
+
+    # https://stackoverflow.com/a/10586169/7086623
```
```suggestion
    # Split the string VICTIM_LIST into an array victim_array: https://stackoverflow.com/a/10586169/7086623
```

tomtseng commented on cpp/evaluate_loop.sh:
```diff
@@ -26,52 +31,63 @@ do
+                if [ "$STEP" -gt "$LAST_STEP" ]; then
+                    # https://stackoverflow.com/questions/12152626/how-can-i-remove-the-extension-of-a-filename-in-a-shell-script
+                    VICTIM_NAME=$(echo "$VICTIM" | cut -f 1 -d '.')
+                    EXTRA_CONFIG="numGamesTotal=100,numGameThreads=10,maxMovesPerGame=300"
```
why these config changes? `maxMovePerGame=300` in particular seems like it will cut a lot of games short

UFO-101 replied:
Oh, oops, I was using for debugging. Removed 👍

AdamGleave reviewed (changes requested):
Agree with Tom's comments (thanks for the review).

One additional question: will this evaluation loop be able to keep up with training? Some concern it could just fall further and further behind especially when evaluating against large enough number of victims. Might be nice to support skipping steps (just evaluate the latest model), or add some logging to warn if it is falling behind (so people can reduce # of victims or increase # of GPUs allocated to job or similar).

UFO-101 committed:
Respond to Tom and Adams comments:
 - Change script to read flags instead of positional args.
 - Improve documentation in usage (--help/-h) message.

UFO-101 commented:
Thanks both for your comments. I've switched to flags, following the style of `launch-job.sh`. There's also better usage info accessible through `-h/--help`.

> One additional question: will this evaluation loop be able to keep up with training?

Hmm, not really sure how to answer or address this right now, as I'm not really familiar with the training code. I actually thought this script was just for evaluation (not training). So I definitely can look into this, but it will take a bit more time.

AdamGleave reviewed (approved):
LGTM

AdamGleave commented:
> Thanks both for your comments. I've switched to flags, following the style of `launch-job.sh`. There's also better usage info accessible through `-h/--help`.

Thanks for making the changes, Joseph!
> Hmm, not really sure how to answer or address this right now, as I'm not really familiar with the training code. I actually thought this script was just for evaluation (not training). So I definitely can look into this, but it will take a bit more time.

OK no need to look into this now, @tomtseng can just monitor the situation in his upcoming training runs and we can parallelize it if needed.

For context, it's not used _for training_ (i.e. it has no effect on the model learned), but it is used for real-time evaluation _of the training process_. So, if evaluation is slower than training, we'll have increasingly delayed metrics.

tomtseng commented:
I looked at our coldstart training run just now `ls -lv /nas/ucb/k8/go-attack/victimplay/ttseng-avoid-pass-alive-coldstart-39-20221025-175949/eval/logs`, and evaluation fell behind training significantly.
* We trained adversary models faster than we evaluated then. Up to the end of October, training would output a model every <1 hour whereas evaluation finished every 4 hours. Training hit 213mil steps and evaluation hit 33mil steps by October 31.
* Then I restarted the coldstart training job to fix a bug after the curriculum hit cp505-v2. Evaluation lost its progress here because it started evaluating the adversary models from the very beginning, and it evaluated the models on the most recent and more expensive victim, cp505. Evaluation hit 21mil steps
* I think Adam restarted the coldstart training job yesterday, so now evaluation lost its progress again and is re-evaluating every adversary model on the most recent victim, cp505-v262144.

So there are a couple issues to resolve if we have extra time:
* Evaluation indeed can fall behind training even in a typical experiment.
* Restarting makes us evaluate all adversaries from the very beginning.

Maybe the answer is to have the eval loop always evaluate the latest adversary it sees rather than having it evaluate every adversary in ascending order.

AdamGleave commented:
> So there are a couple issues to resolve if we have extra time:
> 
> * Evaluation indeed can fall behind training even in a typical experiment.
> * Restarting makes us evaluate all adversaries from the very beginning.
> 
> Maybe the answer is to have the eval loop always evaluate the latest adversary it sees rather than having it evaluate every adversary in ascending order.

Oh, these are pretty major issues.

I do think always evaluating the latest adversary is a reasonable solution for the purpose of statistics during training. Would you be OK making a PR to do that?

We could have a more sophisticated solution that adds more parallelism to evaluation, and saves the last evaluated step to disk to allow resumes. But I think that'd only be worth it if were going to use these eval stats in final publication results rather than just for internal monitoring. But right now we have a separate workflow for that which seems like it's working OK, so probably no need to make `evaluate_loop.sh` more complex.
